From f4a36d9311f0e1f2f6762475100c8e52df6a7153 Mon Sep 17 00:00:00 2001
From: Gary Verhaegen <gary.verhaegen@digitalasset.com>
Date: Thu, 11 Aug 2022 15:25:40 +0200
Subject: [PATCH] m1

---
 scripts/packages/bazel.sh | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/scripts/packages/bazel.sh b/scripts/packages/bazel.sh
index 03b7417369..1f8fa63244 100755
--- a/scripts/packages/bazel.sh
+++ b/scripts/packages/bazel.sh
@@ -16,6 +16,9 @@
 
 set -eu
 
+echo "HEY!!"
+uname -a
+
 # This is a script which can be installed as your "bazel" binary instead of the
 # real Bazel binary. When called, it tries to determine and run the correct
 # Bazel version for a given project and forwards all arguments to it.
@@ -103,7 +106,8 @@ function get_workspace_root() {
 get_workspace_root
 
 readonly wrapper_dir="$(dirname "$(get_realpath "${BASH_SOURCE[0]}")")"
-readonly os_arch_suffix="$(uname -s | tr '[:upper:]' '[:lower:]')-$(uname -m)"
+readonly os_arch_suffix=darwin-arm64 #"$(uname -s | tr '[:upper:]' '[:lower:]')-$(uname -m)"
+echo os_arch_suffix: \"$os_arch_suffix\"
 
 function get_bazel_version() {
   bazel_version=""
@@ -142,6 +146,7 @@ function get_bazel_version() {
     fi
   fi
 
+  echo "bazel_version: \"$bazel_version\""
   readonly reason
   readonly bazel_version
 }
-- 
2.37.1

