From b18f9b6c6724d04f45fb41261a4a03ffe19f6a5d Mon Sep 17 00:00:00 2001
From: Gary Verhaegen <gary.verhaegen@digitalasset.com>
Date: Thu, 11 Aug 2022 15:37:15 +0200
Subject: [PATCH] m1

---
 scripts/packages/bazel.sh | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/scripts/packages/bazel.sh b/scripts/packages/bazel.sh
index 03b7417369..a0eaa008c6 100755
--- a/scripts/packages/bazel.sh
+++ b/scripts/packages/bazel.sh
@@ -16,6 +16,9 @@
 
 set -eu
 
+echo "HEY!!"
+uname -a
+
 # This is a script which can be installed as your "bazel" binary instead of the
 # real Bazel binary. When called, it tries to determine and run the correct
 # Bazel version for a given project and forwards all arguments to it.
@@ -103,7 +106,8 @@ function get_workspace_root() {
 get_workspace_root
 
 readonly wrapper_dir="$(dirname "$(get_realpath "${BASH_SOURCE[0]}")")"
-readonly os_arch_suffix="$(uname -s | tr '[:upper:]' '[:lower:]')-$(uname -m)"
+readonly os_arch_suffix=darwin-arm64 #"$(uname -s | tr '[:upper:]' '[:lower:]')-$(uname -m)"
+echo os_arch_suffix: \"$os_arch_suffix\"
 
 function get_bazel_version() {
   bazel_version=""
@@ -142,6 +146,7 @@ function get_bazel_version() {
     fi
   fi
 
+  echo "bazel_version: \"$bazel_version\""
   readonly reason
   readonly bazel_version
 }
@@ -166,6 +171,8 @@ if [[ ! -x ${BAZEL_REAL} && -x ${bazel_real_path} ]]; then
   fi
 fi
 
+echo "BAZEL_REAL=\"$BAZEL_REAL\""
+
 # If the repository contains a checked-in executable called tools/bazel, we
 # assume that they know what they're doing and have their own way of versioning
 # Bazel. Thus, we don't have to print our helpful messages or error out in case
-- 
2.37.1

