diff --git a/haskell/cc.bzl b/haskell/cc.bzl
--- a/haskell/cc.bzl
+++ b/haskell/cc.bzl
@@ -119,7 +119,14 @@ def cc_interop_info(ctx, override_cc_toolchain = None):
     hs_toolchain = ctx.toolchains["@rules_haskell//haskell:toolchain"]
     cc_wrapper = hs_toolchain.cc_wrapper
     cc = cc_wrapper.executable.path
     cc_files = depset(transitive = [cc_toolchain.all_files, cc_wrapper.inputs])
     cc_manifests = cc_wrapper.manifests
 
+    if hs_toolchain.is_darwin and not hs_toolchain.is_arm64:
+        linker_flags = [
+            flag
+            for flag in linker_flags
+            if flag != "-fuse-ld=lld"
+        ]
+
     cc_bindir = real_cc_path[:real_cc_path.rfind("/") + 1]
