diff --git a/ghc-lib-gen/src/Ghclibgen.hs b/ghc-lib-gen/src/Ghclibgen.hs
index 3574255..4fee612 100644
--- a/ghc-lib-gen/src/Ghclibgen.hs
+++ b/ghc-lib-gen/src/Ghclibgen.hs
@@ -336,10 +336,7 @@ calcParserModules ghcFlavor = do
       hsSrcIncludes = map ("-i" ++ ) (placeholderModulesDir : hsSrcDirs)
       -- See [Note: GHC now depends on exceptions package].
       cmd = unwords $
-        [ "stack exec" ] ++
-        [ "--package exceptions" | ghcFlavor >= Ghc901 ] ++
-        [ "--stack-yaml hadrian/stack.yaml" ] ++
-        [ "-- ghc"
+        [ "ghc"
         , "-dep-suffix ''"
         , "-dep-makefile .parser-depends"
         , "-M"
@@ -1258,15 +1255,13 @@ generatePrerequisites ghcFlavor = do
   -- If building happy in the next step, the configure it does
   -- requires some versions of alex and happy pre-exist. We make sure
   -- of this in CI.hs.
-  system_ "stack --stack-yaml hadrian/stack.yaml build --only-dependencies"
-  system_ "stack --stack-yaml hadrian/stack.yaml exec -- bash -c ./boot"
-  system_ "stack --stack-yaml hadrian/stack.yaml exec -- bash -c \"./configure --enable-tarballs-autodownload\""
+  system_ "bash -c ./boot"
+  system_ "bash -c \"./configure --enable-tarballs-autodownload\""
   withCurrentDirectory "hadrian" $ do
     -- No need to specify a stack.yaml here, we are in the hadrian
     -- directory itself.
-    system_ "stack build --no-library-profiling"
     system_ $ unwords $
-        [ "stack exec hadrian --"
+        [ "hadrian"
           , "--directory=.."
           , "--build-root=ghc-lib"
         ] ++
