diff --git a/nix/nixpkgs.nix b/nix/nixpkgs.nix
index 8c7eb189783..e0b7e1e5534 100644
--- a/nix/nixpkgs.nix
+++ b/nix/nixpkgs.nix
@@ -38,8 +38,8 @@ let
         # for the only change that actually affects the code in this patch. The rest is tests
         # and/or documentation.
         (pkgs.fetchurl {
-          url = "https://patch-diff.githubusercontent.com/raw/bazelbuild/bazel/pull/8983.patch";
-          sha256 = "1j25bycn9q7536ab3ln6yi6zpzv2b25fwdyxbgnalkpl2dz9idb7";
+          url = "https://github.com/bazelbuild/bazel/commit/0f0a0d58725603cf2f1c175963360b525718a195.patch";
+          sha256 = "1h9sqlcz3nkpppkbk0rilwf7ivhxp9iwircfpm1mbbjwq2pwywqa";
         })
       ];
     });
