diff --git a/google/rpc/BUILD.bazel b/google/rpc/BUILD.bazel
index 8ffa5a49c..eb0a8127b 100644
--- a/google/rpc/BUILD.bazel
+++ b/google/rpc/BUILD.bazel
@@ -1,5 +1,7 @@
 load("@rules_proto//proto:defs.bzl", "proto_library")
 
+exports_files(["status.proto", "error_details.proto"])
+
 # This is an API workspace, having public visibility by default makes perfect sense.
 package(default_visibility = ["//visibility:public"])
 
