diff --git a/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java b/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java
index 34a5d9bebd..380b7236e5 100644
--- a/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java
@@ -30,6 +30,7 @@ import com.google.devtools.build.lib.events.Event;
 import com.google.devtools.build.lib.events.Reporter;
 import com.google.devtools.build.lib.exec.ExecutorBuilder;
 import com.google.devtools.build.lib.remote.logging.LoggingInterceptor;
+import com.google.devtools.build.lib.remote.RemoteRetrier.ExponentialBackoff;
 import com.google.devtools.build.lib.remote.util.DigestUtil;
 import com.google.devtools.build.lib.remote.util.TracingMetadataUtils;
 import com.google.devtools.build.lib.runtime.BlazeModule;
@@ -250,8 +251,8 @@ public final class RemoteModule extends BlazeModule {
       if (enableBlobStoreCache) {
         Retrier retrier =
             new Retrier(
-                () -> Retrier.RETRIES_DISABLED,
-                (e) -> false,
+                () -> new ExponentialBackoff(remoteOptions),
+                (e) -> true,
                 retryScheduler,
                 Retrier.ALLOW_ALL_CALLS);
         executeRetrier = null;
