diff --git a/haskell/private/actions/package.bzl b/haskell/private/actions/package.bzl
index f46f56bc..bfcc9500 100644
--- a/haskell/private/actions/package.bzl
+++ b/haskell/private/actions/package.bzl
@@ -107,7 +107,12 @@ def package(
         "library-dirs": ["${pkgroot}"] + extra_lib_dirs,
         "dynamic-library-dirs": ["${pkgroot}"] + extra_dynamic_lib_dirs,
         "hs-libraries": [pkg_id.library_name(hs, my_pkg_id)] if has_hs_library else [],
-        "extra-libraries": extra_libs,
+        "extra-libraries": extra_libs + ([
+            "stdc++",
+            "crypt32",
+            "shlwapi",
+            "ws2_32",
+        ] if hs.toolchain.is_windows else []),
         "depends": hs.package_ids,
     })
 
