diff --git a/ghcide.cabal b/ghcide.cabal
index d11d000..2ee2c0a 100644
--- a/ghcide.cabal
+++ b/ghcide.cabal
@@ -96,6 +96,7 @@ library
     include-dirs:
         include
     exposed-modules:
+        Development.IDE.Core.Compile
         Development.IDE.Core.FileStore
         Development.IDE.Core.OfInterest
         Development.IDE.Core.PositionMapping
@@ -103,7 +104,10 @@ library
         Development.IDE.Core.RuleTypes
         Development.IDE.Core.Service
         Development.IDE.Core.Shake
+        Development.IDE.GHC.Compat
+        Development.IDE.GHC.Error
         Development.IDE.GHC.Util
+        Development.IDE.GHC.Warnings
         Development.IDE.Import.DependencyInformation
         Development.IDE.LSP.LanguageServer
         Development.IDE.LSP.Protocol
@@ -114,13 +118,9 @@ library
         Development.IDE.Types.Options
     other-modules:
         Development.IDE.Core.Debouncer
-        Development.IDE.Core.Compile
         Development.IDE.Core.Preprocessor
-        Development.IDE.GHC.Compat
         Development.IDE.GHC.CPP
-        Development.IDE.GHC.Error
         Development.IDE.GHC.Orphans
-        Development.IDE.GHC.Warnings
         Development.IDE.Import.FindImports
         Development.IDE.LSP.CodeAction
         Development.IDE.LSP.Definition
