diff --git a/haskell/repl.bzl b/haskell/repl.bzl
index db072175..22108860 100644
--- a/haskell/repl.bzl
+++ b/haskell/repl.bzl
@@ -415,6 +415,12 @@ def _create_hie_bios(hs, posix, ctx, repl_info):
     args.extend(hs.toolchain.compiler_flags)
     args.extend(repl_info.load_info.compiler_flags)
 
+    # Add import directories.
+    # Note, src_strip_prefix is deprecated. However, for now ghcide depends on
+    # `-i` flags to find source files to modules.
+    for import_dir in repl_info.load_info.import_dirs.to_list():
+        args.append("-i" + (import_dir if import_dir else "."))
+
     args_file = ctx.actions.declare_file(".%s.hie-bios" % ctx.label.name)
     args_link = ctx.actions.declare_file("%s@hie-bios" % ctx.label.name)
     ctx.actions.write(args_file, "\n".join(args))
