diff --git a/src/HIE/Bios/Cradle.hs b/src/HIE/Bios/Cradle.hs
index 291a706..7a37977 100644
--- a/src/HIE/Bios/Cradle.hs
+++ b/src/HIE/Bios/Cradle.hs
@@ -125,10 +125,10 @@ cabalCradle wdir mc = do
   }
 
 cabalWrapper :: String
-cabalWrapper = $(embedStringFile "wrappers/cabal")
+cabalWrapper = $(embedStringFile "external/haskell_hie__bios/wrappers/cabal")
 
 cabalWrapperBat :: String
-cabalWrapperBat = $(embedStringFile "wrappers/cabal.bat")
+cabalWrapperBat = $(embedStringFile "external/haskell_hie__bios/wrappers/cabal.bat")
 
 cabalAction :: FilePath -> Maybe String -> FilePath -> IO (ExitCode, String, [String])
 cabalAction work_dir mc _fp = do
@@ -179,7 +179,7 @@ stackCradle wdir =
 
 -- Same wrapper works as with cabal
 stackWrapper :: String
-stackWrapper = $(embedStringFile "wrappers/cabal")
+stackWrapper = $(embedStringFile "external/haskell_hie__bios/wrappers/cabal")
 
 stackAction :: FilePath -> FilePath -> IO (ExitCode, String, [String])
 stackAction work_dir fp = do
@@ -228,7 +228,7 @@ rulesHaskellCradle wdir = do
 
 
 bazelCommand :: String
-bazelCommand = $(embedStringFile "wrappers/bazel")
+bazelCommand = $(embedStringFile "external/haskell_hie__bios/wrappers/bazel")
 
 rulesHaskellAction :: FilePath -> FilePath -> IO (ExitCode, String, [String])
 rulesHaskellAction work_dir fp = do
