diff --git a/haskell/private/actions/compile.bzl b/haskell/private/actions/compile.bzl
index 7d882078..b31790fc 100644
--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -260,7 +260,7 @@ def _compilation_defaults(hs, cc, java, posix, dep_info, plugin_dep_info, srcs,
 
     # Compilation mode.  Allow rule-supplied compiler flags to override it.
     if hs.mode == "opt":
-        args.add("-O2")
+        args.add("-O")
 
     args.add("-static")
     if with_profiling:
