diff --git a/tools/runfiles/src/Bazel/Runfiles.hs b/tools/runfiles/src/Bazel/Runfiles.hs
index 96b3aeb..665e0ee 100644
--- a/tools/runfiles/src/Bazel/Runfiles.hs
+++ b/tools/runfiles/src/Bazel/Runfiles.hs
@@ -189,4 +189,4 @@ parseManifest = map parseLine . lines
   where
     parseLine l =
         let (key, value) = span (/= ' ') l in
-        (key, dropWhile (== ' ') value)
+        (normalize key, normalize $ dropWhile (== ' ') value)
