diff --git a/haskell/ghc_bindist.bzl b/haskell/ghc_bindist.bzl
index fd516855..6b6bbe96 100644
--- a/haskell/ghc_bindist.bzl
+++ b/haskell/ghc_bindist.bzl
@@ -294,6 +294,20 @@ GHC_BINDIST = \
                 "e9175a276504c3390a5e0084954e6997d56078737dbe7158049518892cf6bfb2",
             ),
         },
+        "8.10.7": {
+            "darwin_amd64": (
+                "https://downloads.haskell.org/~ghc/8.10.7/ghc-8.10.7-x86_64-apple-darwin.tar.xz",
+                "287db0f9c338c9f53123bfa8731b0996803ee50f6ee847fe388092e5e5132047",
+            ),
+            "linux_amd64": (
+                "https://downloads.haskell.org/~ghc/8.10.7/ghc-8.10.7-x86_64-deb9-linux.tar.xz",
+                "ced9870ea351af64fb48274b81a664cdb6a9266775f1598a79cbb6fdd5770a23",
+            ),
+            "windows_amd64": (
+                "https://downloads.haskell.org/~ghc/8.10.7/ghc-8.10.7-x86_64-unknown-mingw32.tar.xz",
+                "b6515b0ea3f7a6e34d92e7fcd0c1fef50d6030fe8f46883000185289a4b8ea9a",
+            ),
+        },
         "9.0.1": {
             "darwin_amd64": (
                 "https://downloads.haskell.org/~ghc/9.0.1/ghc-9.0.1-x86_64-apple-darwin.tar.xz",
