diff --git a/core/grpc-haskell-core.cabal b/core/grpc-haskell-core.cabal
index 8c67ce9..8ffd432 100644
--- a/core/grpc-haskell-core.cabal
+++ b/core/grpc-haskell-core.cabal
@@ -10,6 +10,7 @@ copyright:           Copyright 2016 Awake Networks
 category:            Network
 build-type:          Simple
 cabal-version:       >=1.10
+extra-source-files:  include/grpc_haskell.h
 
 Flag Debug
   Description: Adds debug logging.
