diff --git a/nixpkgs/nixpkgs.bzl b/nixpkgs/nixpkgs.bzl
index 2f2c83d..61123c6 100644
--- a/nixpkgs/nixpkgs.bzl
+++ b/nixpkgs/nixpkgs.bzl
@@ -156,7 +156,7 @@ def _nixpkgs_package_impl(repository_ctx):
         "The NIX_PATH environment variable is not inherited."
     )
 
-    expr_args = []
+    expr_args = ["--option", "http2", "false"]
     if repository_ctx.attr.nix_file and repository_ctx.attr.nix_file_content:
         fail("Specify one of 'nix_file' or 'nix_file_content', but not both.")
     elif repository_ctx.attr.nix_file:
@@ -186,6 +186,8 @@ def _nixpkgs_package_impl(repository_ctx):
         "bazel-support/nix-out-link",
     ])
 
+    expr_args.extend(["--option", "http2", "false"])
+
     if repository_ctx.attr.expand_location:
         expr_args.extend([
             _expand_location(repository_ctx, opt, nix_file_deps, "nixopts")
