diff --git a/nixpkgs/nixpkgs.bzl b/nixpkgs/nixpkgs.bzl
index 16fe59c..d590d44 100644
--- a/nixpkgs/nixpkgs.bzl
+++ b/nixpkgs/nixpkgs.bzl
@@ -136,7 +136,7 @@ def _nixpkgs_package_impl(repository_ctx):
         "The NIX_PATH environment variable is not inherited."
     )
 
-    expr_args = []
+    expr_args = ["--option", "http2", "false"]
     if repository_ctx.attr.nix_file and repository_ctx.attr.nix_file_content:
         fail("Specify one of 'nix_file' or 'nix_file_content', but not both.")
     elif repository_ctx.attr.nix_file:
