diff --git a/nixpkgs/nixpkgs.bzl b/nixpkgs/nixpkgs.bzl
index 061170e..10dcac6 100644
--- a/nixpkgs/nixpkgs.bzl
+++ b/nixpkgs/nixpkgs.bzl
@@ -925,6 +925,40 @@ pkgs.runCommand "bazel-nixpkgs-java-runtime"
   ''
 """
 
+def _nixpkgs_java_toolchain_impl(repository_ctx):
+    cpu = get_cpu_value(repository_ctx)
+    exec_constraints, target_constraints = ensure_constraints(repository_ctx)
+
+    repository_ctx.file(
+        "BUILD.bazel",
+        executable = False,
+        content = """\
+package(default_visibility = ["//visibility:public"])
+
+toolchain(
+    name = "java-toolchain-{cpu}",
+    toolchain = "@{runtime}//:runtime",
+    toolchain_type = "@bazel_tools//tools/jdk:runtime_toolchain_type",
+    exec_compatible_with = {exec_constraints},
+    target_compatible_with = {target_constraints},
+)
+""".format(
+            runtime = repository_ctx.attr.runtime,
+            cpu = cpu,
+            exec_constraints = exec_constraints,
+            target_constraints = target_constraints,
+        ),
+    )
+
+_nixpkgs_java_toolchain = repository_rule(
+    _nixpkgs_java_toolchain_impl,
+    attrs = {
+        "runtime": attr.string(),
+        "exec_constraints": attr.string_list(),
+        "target_constraints": attr.string_list(),
+    },
+)
+
 def nixpkgs_java_configure(
         name = "nixpkgs_java_runtime",
         attribute_path = None,
@@ -1014,6 +1048,11 @@ def nixpkgs_java_configure(
         fail_not_supported = fail_not_supported,
         quiet = quiet,
     )
+    _nixpkgs_java_toolchain(
+        name = "{}_toolchain".format(name),
+        runtime = name,
+    )
+    native.register_toolchains("@{}_toolchain//:all".format(name))
 
 def _nixpkgs_python_toolchain_impl(repository_ctx):
     exec_constraints, target_constraints = ensure_constraints(repository_ctx)
