# See https://github.com/protocolbuffers/protobuf/issues/8049
diff --git a/src/google/protobuf/port_def.inc b/src/google/protobuf/port_def.inc
index 320e888ec..e24b8808c 100644
--- a/src/google/protobuf/port_def.inc
+++ b/src/google/protobuf/port_def.inc
@@ -560,10 +560,10 @@
 #define PROTOBUF_CONSTINIT
 #endif
 
-// Some constructors can't be constexpr under MSVC, but given that MSVC will not
+// Some constructors can't be constexpr under Windows, but given that Windows will not
 // do constant initialization of globals anyway we can omit `constexpr` from
 // them. These constructors are marked with PROTOBUF_MAYBE_CONSTEXPR
-#if defined(_MSC_VER)
+#if defined(_WIN32)
 #define PROTOBUF_MAYBE_CONSTEXPR
 #else
 #define PROTOBUF_MAYBE_CONSTEXPR constexpr
