diff --git a/core/util.bzl b/core/util.bzl
index eff10e7..b90080a 100644
--- a/core/util.bzl
+++ b/core/util.bzl
@@ -110,10 +110,13 @@ def default_constraints(repository_ctx):
+    print("cpu: ", cpu_value)
     cpu = {
         "darwin": "@platforms//cpu:x86_64",
         "darwin_arm64": "@platforms//cpu:arm64",
+        "aarch64": "@platforms//cpu:arm64",
     }.get(cpu_value, "@platforms//cpu:x86_64")
     os = {
         "darwin": "@platforms//os:osx",
         "darwin_arm64": "@platforms//os:osx",
+        "x64_windows": "@platforms//os:windows",
     }.get(cpu_value, "@platforms//os:linux")
     return [cpu, os]
 

