diff --git a/src/java/io/bazel/rulesscala/exe/LauncherFileWriter.java b/src/java/io/bazel/rulesscala/exe/LauncherFileWriter.java
index 9fd8f9f..c14fb41 100644
--- a/src/java/io/bazel/rulesscala/exe/LauncherFileWriter.java
+++ b/src/java/io/bazel/rulesscala/exe/LauncherFileWriter.java
@@ -30,7 +30,7 @@ public class LauncherFileWriter {
             .addKeyValuePair("binary_type", "Java")
             .addKeyValuePair("workspace_name", workspaceName)
             .addKeyValuePair("symlink_runfiles_enabled", "0")
-            .addKeyValuePair("java_bin_path", workspaceName + "/" + javaBinPath)
+            .addKeyValuePair("java_bin_path", fullJavaBinPath(workspaceName, Paths.get(javaBinPath)).toString())
             .addKeyValuePair("jar_bin_path", jarBinPath)
             .addKeyValuePair("java_start_class", javaStartClass)
             .addKeyValuePair("classpath", classpath)
@@ -52,4 +52,15 @@ public class LauncherFileWriter {
       out.flush();
     }
   }
+
+  private static Path fullJavaBinPath(String workspaceName, Path javaBinPath) {
+    if (javaBinPath.isAbsolute()) {
+      return javaBinPath;
+    } else if (javaBinPath.startsWith(Paths.get("external"))) {
+      // Paths under `external/` already have a workspace name.
+      return javaBinPath;
+    } else {
+      return Paths.get(workspaceName).resolve(javaBinPath);
+    }
+  }
 }
