diff --git a/haskell/private/ls_modules.py b/haskell/private/ls_modules.py
index bb4d616..9e54b7f 100755
--- a/haskell/private/ls_modules.py
+++ b/haskell/private/ls_modules.py
@@ -26,6 +26,8 @@ hidden_modules_file = sys.argv[4]
 reexported_modules_file = sys.argv[5]
 results_file = sys.argv[6]
 
+print("ARGS: {}".format(sys.argv))
+
 with io.open(global_pkg_db_dump, "r", encoding='utf8') as f:
     names = [line.split()[1] for line in f if line.startswith("name:")]
     f.seek(0)
@@ -85,12 +87,20 @@ On Windows you may need to enable long file path support:
     """.strip().format(e), file=sys.stderr)
     exit(1)
 
+interface_files_debug = (
+    os.path.join(path, f)
+    for path, dirs, files in os.walk(root, onerror=handle_walk_error)
+    for f in fnmatch.filter(files, '*.p_hi' if with_profiling else '*.hi')
+)
+
 interface_files = (
     os.path.join(path, f)
     for path, dirs, files in os.walk(root, onerror=handle_walk_error)
     for f in fnmatch.filter(files, '*.p_hi' if with_profiling else '*.hi')
 )
 
+print("MODULES: {}".format([f for f in interface_files_debug]))
+
 modules = (
     # replace directory separators by . to generate module names
     # / and \ are respectively the separators for unix (linux / darwin) and windows systems
