diff --git a/jmh/jmh.bzl b/jmh/jmh.bzl
index de45c9a..3fd6c63 100644
--- a/jmh/jmh.bzl
+++ b/jmh/jmh.bzl
@@ -56,8 +56,8 @@ def jmh_repositories(maven_servers = _default_maven_server_urls()):
     )
     _scala_maven_import_external(
         name = "io_bazel_rules_scala_net_sf_jopt_simple_jopt_simple",
-        artifact = "net.sf.jopt-simple:jopt-simple:5.0.3",
-        artifact_sha256 = "6f45c00908265947c39221035250024f2caec9a15c1c8cf553ebeecee289f342",
+        artifact = "net.sf.jopt-simple:jopt-simple:4.6",
+        artifact_sha256 = "3fcfbe3203c2ea521bf7640484fd35d6303186ea2e08e72f032d640ca067ffda",
         licenses = ["notice"],
         server_urls = maven_servers,
     )
