diff --git a/scala/private/phases/phase_default_info.bzl b/scala/private/phases/phase_default_info.bzl
index 53f0568..ad3678f 100644
--- a/scala/private/phases/phase_default_info.bzl
+++ b/scala/private/phases/phase_default_info.bzl
@@ -7,6 +7,7 @@
 def phase_default_info(ctx, p):
     executable = None
     files = []
+    direct = None
     runfiles = []
 
     phase_names = dir(p)
@@ -18,6 +19,7 @@ def phase_default_info(ctx, p):
         if hasattr(phase, "executable"):
             if executable == None:
                 executable = phase.executable
+                direct = [executable]
             else:
                 fail("only one executable may be provided")
 
@@ -31,11 +33,11 @@ def phase_default_info(ctx, p):
         external_providers = {
             "DefaultInfo": DefaultInfo(
                 executable = executable,
-                files = depset(transitive = files),
+                files = depset(direct = direct, transitive = files),
                 # TODO:
                 # Per Bazel documentation, we should avoid using collect_data. The core phases need to be updated
                 # before we can make the adjustment.
-                runfiles = ctx.runfiles(transitive_files = depset(transitive = runfiles), collect_data = True),
+                runfiles = ctx.runfiles(transitive_files = depset(direct = direct, transitive = runfiles), collect_data = True),
             ),
         },
     )
