--- src/System/PosixCompat/Files.hsc    2018-08-23 02:44:16.000000000 +0200
+++ src/System/PosixCompat/Files.hsc    2019-04-15 13:41:46.371761110 +0200
@@ -392,11 +392,7 @@
 -- Renaming

 rename :: FilePath -> FilePath -> IO ()
-#if MIN_VERSION_Win32(2, 6, 0)
 rename name1 name2 = moveFileEx name1 (Just name2) mOVEFILE_REPLACE_EXISTING
-#else
-rename name1 name2 = moveFileEx name1 name2 mOVEFILE_REPLACE_EXISTING
-#endif

 -- -----------------------------------------------------------------------------
 -- chown()
