diff --git a/bazel/grpc_deps.bzl b/bazel/grpc_deps.bzl
index cce2f88fe8..ad7ecf0286 100644
--- a/bazel/grpc_deps.bzl
+++ b/bazel/grpc_deps.bzl
@@ -226,10 +226,10 @@ def grpc_deps():
         )
 
     if "build_bazel_rules_apple" not in native.existing_rules():
-        git_repository(
+        http_archive(
             name = "build_bazel_rules_apple",
-            remote = "https://github.com/bazelbuild/rules_apple.git",
-            tag = "0.17.2",
+            urls = ["https://github.com/bazelbuild/rules_apple/releases/download/0.17.2/rules_apple.0.17.2.tar.gz"],
+            sha256 = "6efdde60c91724a2be7f89b0c0a64f01138a45e63ba5add2dca2645d981d23a1",
         )
     
 # TODO: move some dependencies from "grpc_deps" here?
