diff --git a/haskell/private/actions/compile.bzl b/haskell/private/actions/compile.bzl
index 1dc77cee..60cb70e2 100644
--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -214,7 +214,7 @@ def _compilation_defaults(
 
     # Compilation mode.  Allow rule-supplied compiler flags to override it.
     if hs.mode == "opt":
-        args.add("-O2")
+        args.add("-O")
 
     args.add("-static")
     if with_profiling:
