diff --git a/src/HIE/Bios/Cradle.hs b/src/HIE/Bios/Cradle.hs
index 8ca8251..672255e 100644
--- a/src/HIE/Bios/Cradle.hs
+++ b/src/HIE/Bios/Cradle.hs
@@ -146,10 +146,10 @@ cabalCradle wdir mc =
   }
 
 cabalWrapper :: String
-cabalWrapper = $(embedStringFile "wrappers/cabal")
+cabalWrapper = $(embedStringFile "external/haskell_hie__bios/wrappers/cabal")
 
 cabalWrapperHs :: String
-cabalWrapperHs = $(embedStringFile "wrappers/cabal.hs")
+cabalWrapperHs = $(embedStringFile "external/haskell_hie__bios/wrappers/cabal.hs")
 
 processCabalWrapperArgs :: String -> Maybe [String]
 processCabalWrapperArgs args =
@@ -221,7 +221,7 @@ stackCradle wdir =
 
 -- Same wrapper works as with cabal
 stackWrapper :: String
-stackWrapper = $(embedStringFile "wrappers/cabal")
+stackWrapper = $(embedStringFile "external/haskell_hie__bios/wrappers/cabal")
 
 stackAction :: FilePath -> FilePath -> IO (ExitCode, String, [String])
 stackAction work_dir fp = do
@@ -272,7 +272,7 @@ rulesHaskellCradle wdir =
 
 
 bazelCommand :: String
-bazelCommand = $(embedStringFile "wrappers/bazel")
+bazelCommand = $(embedStringFile "external/haskell_hie__bios/wrappers/bazel")
 
 rulesHaskellAction :: FilePath -> FilePath -> IO (ExitCode, String, [String])
 rulesHaskellAction work_dir fp = do
