diff --git a/haskell/private/actions/package.bzl b/haskell/private/actions/package.bzl
index 842b9eee..2c0e31a5 100644
--- a/haskell/private/actions/package.bzl
+++ b/haskell/private/actions/package.bzl
@@ -102,7 +102,13 @@ def package(
         "import-dirs": [import_dir] if has_iface else [],
         "library-dirs": [pkgroot_lib_path] + extra_lib_dirs,
         "dynamic-library-dirs": [pkgroot_lib_path] + extra_dynamic_lib_dirs,
-        "extra-libraries": extra_libs,
+        "extra-libraries": extra_libs + ([
+            "stdc++",
+            "crypt32",
+            "shlwapi",
+            "ws2_32",
+            "bcrypt",
+        ] if hs.toolchain.is_windows else []),
         "depends": hs.package_ids,
         # TODO[AH] Add haskell_module modules
         "exposed-modules": exposed_modules,
