diff --git a/toolchains/java/default_java_toolchain.bzl b/toolchains/java/default_java_toolchain.bzl
index 1919433..6cf40bc 100644
--- a/toolchains/java/default_java_toolchain.bzl
+++ b/toolchains/java/default_java_toolchain.bzl
@@ -42,8 +42,8 @@ BASE_JDK9_JVM_OPTS = [
 
 JDK9_JVM_OPTS = BASE_JDK9_JVM_OPTS + [
     # override the javac in the JDK.
-    "--patch-module=java.compiler=$(location @remote_java_tools//:java_compiler_jar)",
-    "--patch-module=jdk.compiler=$(location @remote_java_tools//:jdk_compiler_jar)",
+    #"--patch-module=java.compiler=$(location @remote_java_tools//:java_compiler_jar)",
+    #"--patch-module=jdk.compiler=$(location @remote_java_tools//:jdk_compiler_jar)",
 ]
 
 DEFAULT_JAVACOPTS = [
-- 
2.34.1

