diff --git a/src/opencensus/proto/trace/v1/BUILD.bazel b/src/opencensus/proto/trace/v1/BUILD.bazel
index 0beb6d8..b11f455 100644
--- a/src/opencensus/proto/trace/v1/BUILD.bazel
+++ b/src/opencensus/proto/trace/v1/BUILD.bazel
@@ -77,7 +77,6 @@ py_proto_library(
     srcs = ["trace.proto"],
     deps = [
         "//opencensus/proto/resource/v1:resource_proto_py",
-        "@com_google_protobuf//:protobuf_python",
     ],
 )
 
