diff --git a/haskell/private/ghc_wrapper.sh b/haskell/private/ghc_wrapper.sh
index 0030bf12..c04d7c3b 100755
--- a/haskell/private/ghc_wrapper.sh
+++ b/haskell/private/ghc_wrapper.sh
@@ -35,8 +35,20 @@ else
 
     # We use the information gathered from those 2 files to compile the Haskell file.
 
-    "${compile_flags[@]}" "${extra_args[@]}" 2>&1 \
-        | drop_loaded_and_warning >&2
+    if [ "$(uname -s)" = "Darwin" ] && [ "$(uname -m)" = "arm64" ]; then
+      prior_stack_limit=$(ulimit -s)
+      ulimit -s 64000
+
+      echo "Running on Apple Silicon (arm64). Changing stack limit from $prior_stack_limit to 64000."
+      if "${compile_flags[@]}" "${extra_args[@]}" 2>&1 | drop_loaded_and_warning >&2; then
+        build_succeeded=true
+        break
+      fi
+      ulimit -s $prior_stack_limit
+    else
+      "${compile_flags[@]}" "${extra_args[@]}" 2>&1 \
+          | drop_loaded_and_warning >&2
+    fi
 
     if [ "$MUST_EXTRACT_ABI" = "true" ]
     then
